File: src/lib/tours/routeVisibility.test.ts
/**
 * npx tsx src/lib/tours/routeVisibility.test.ts
 */
import assert from "node:assert/strict";
import type { SavedRoute } from "../../types/route";
import {
  allowsStimmen,
  filterSavedByVisibility,
  isShared,
  mayContributeRideTrack,
  mayContributeSavedGeometry,
  shareHonesty,
  shareableRouteIds,
  stimmenTourIdOf,
  visibilityOf,
  visibleInPublicExplore,
} from "./routeVisibility";

function route(partial: Partial<SavedRoute> & Pick<SavedRoute, "id">): SavedRoute {
  return {
    name: partial.name ?? "T",
    distanceKm: 10,
    elevationM: 100,
    durationMin: 40,
    savedAt: "2026-08-15T00:00:00.000Z",
    source: partial.source ?? "import",
    ...partial,
  };
}

assert.equal(visibilityOf({}), "private");
assert.equal(visibilityOf({ visibility: undefined }), "private");
assert.equal(visibilityOf({ visibility: "private" }), "private");
assert.equal(visibilityOf({ visibility: "shared" }), "shared");
assert.equal(isShared(route({ id: "gpx-1" })), false);

const priv = route({ id: "gpx-1", source: "import" });
const shared = route({ id: "gpx-2", source: "import", visibility: "shared" });
const catalog = route({
  id: "r-bodensee-road",
  source: "suggestion",
  catalogTourId: "r-bodensee-road",
});

assert.equal(visibleInPublicExplore(priv), false);
assert.equal(visibleInPublicExplore(shared), true);
assert.equal(mayContributeSavedGeometry(priv), false);
assert.equal(mayContributeSavedGeometry(shared), true);
assert.equal(mayContributeRideTrack(undefined), true);
assert.equal(mayContributeRideTrack(priv), false);
assert.equal(mayContributeRideTrack(shared), true);
assert.equal(mayContributeRideTrack(catalog), true);

assert.equal(allowsStimmen(priv), false);
assert.equal(stimmenTourIdOf(priv), null);
assert.equal(allowsStimmen(shared), true);
assert.equal(stimmenTourIdOf(shared), "gpx-2");
assert.equal(allowsStimmen(catalog), true);
assert.equal(stimmenTourIdOf(catalog), "r-bodensee-road");

const recorded = route({ id: "recorded-abc", source: "import" });
assert.equal(allowsStimmen(recorded), false);
assert.equal(stimmenTourIdOf(recorded), null);

const list = [priv, shared, catalog];
assert.equal(filterSavedByVisibility(list, "all_mine").length, 3);
assert.deepEqual(
  filterSavedByVisibility(list, "private").map((r) => r.id),
  ["gpx-1", "r-bodensee-road"]
);
assert.deepEqual(
  filterSavedByVisibility(list, "shared").map((r) => r.id),
  ["gpx-2"]
);

assert.deepEqual(
  shareableRouteIds(["gpx-1", "gpx-2", "r-bodensee-road"], list),
  ["gpx-2", "r-bodensee-road"]
);

const honesty = shareHonesty(
  route({
    id: "gpx-track",
    visibility: "shared",
    geometry: { type: "LineString", coordinates: [[8.6, 49.4], [8.7, 49.5]] },
  })
);
assert.match(honesty, /Server/);
assert.match(honesty, /eingeloggt/);

console.log("routeVisibility.test.ts OK");

File: mobile/test/route_visibility_test.dart
import 'package:aetherride_mobile/domain/routing/tour_filters.dart';
import 'package:aetherride_mobile/domain/saved_route.dart';
import 'package:aetherride_mobile/domain/saved_route_note.dart';
import 'package:aetherride_mobile/domain/tours/route_visibility.dart';
import 'package:flutter_test/flutter_test.dart';

SavedRouteEntry _route(String id) => SavedRouteEntry(
      id: id,
      name: id,
      distanceKm: 10,
      elevationM: 100,
      durationMin: 40,
      savedAt: DateTime.utc(2026, 8, 15),
      source: 'import',
    );

void main() {
  test('default visibility is private', () {
    expect(RouteVisibility.visibilityOf(null), RouteVisibility.private);
    expect(
      RouteVisibility.visibilityOf(SavedRouteMeta.empty),
      RouteVisibility.private,
    );
    expect(
      RouteVisibility.visibilityOf(const SavedRouteMeta()),
      RouteVisibility.private,
    );
    expect(
      SavedRouteMeta.fromJson({}).visibility,
      RouteVisibility.private,
    );
    expect(SavedRouteMeta.fromJson({}).toJson().containsKey('visibility'), isFalse);
  });

  test('shared persists, private stays default', () {
    const shared = SavedRouteMeta(visibility: RouteVisibility.shared);
    expect(shared.isEmpty, isFalse);
    expect(shared.toJson()['visibility'], 'shared');
    expect(RouteVisibility.isShared(shared), isTrue);
    expect(RouteVisibility.visibleInPublicExplore(shared), isTrue);
    expect(RouteVisibility.mayContributeSavedGeometry(shared), isTrue);
    expect(
      RouteVisibility.mayContributeSavedGeometry(SavedRouteMeta.empty),
      isFalse,
    );
  });

  test('Stimmen only for catalog or shared', () {
    expect(RouteVisibility.allowsStimmen('gpx-1'), isFalse);
    expect(RouteVisibility.stimmenTourIdOf('gpx-1'), isNull);
    expect(
      RouteVisibility.allowsStimmen(
        'gpx-2',
        const SavedRouteMeta(visibility: RouteVisibility.shared),
      ),
      isTrue,
    );
    expect(
      RouteVisibility.stimmenTourIdOf(
        'gpx-2',
        const SavedRouteMeta(visibility: RouteVisibility.shared),
      ),
      'gpx-2',
    );
    expect(
      RouteVisibility.stimmenTourIdOf(
        'saved-x',
        const SavedRouteMeta(catalogTourId: 'r-bodensee-road'),
      ),
      'r-bodensee-road',
    );
    expect(RouteVisibility.allowsStimmen('recorded-abc'), isFalse);
    expect(RouteVisibility.stimmenTourIdOf('recorded-abc'), isNull);
  });

  test('filter chips respect visibility', () {
    final routes = [_route('gpx-1'), _route('gpx-2'), _route('r-bodensee-road')];
    final metas = {
      'gpx-2': const SavedRouteMeta(visibility: RouteVisibility.shared),
      'r-bodensee-road':
          const SavedRouteMeta(catalogTourId: 'r-bodensee-road'),
    };
    expect(
      RouteVisibility.filter(routes, TourVisibilityKey.allMine, metas).length,
      3,
    );
    expect(
      RouteVisibility.filter(routes, TourVisibilityKey.privateOnly, metas)
          .map((r) => r.id),
      ['gpx-1', 'r-bodensee-road'],
    );
    expect(
      RouteVisibility.filter(routes, TourVisibilityKey.sharedOnly, metas)
          .map((r) => r.id),
      ['gpx-2'],
    );
    expect(
      TourFilters.visibilityMatches(null, TourVisibilityKey.privateOnly),
      isTrue,
    );
    expect(
      TourFilters.visibilityMatches('shared', TourVisibilityKey.sharedOnly),
      isTrue,
    );
  });

  test('heatmap ride skip for private GPX, allow freeride and catalog', () {
    expect(RouteVisibility.mayContributeRide(null, null), isTrue);
    expect(
      RouteVisibility.mayContributeRide('gpx-1', SavedRouteMeta.empty),
      isFalse,
    );
    expect(
      RouteVisibility.mayContributeRide(
        'gpx-2',
        const SavedRouteMeta(visibility: RouteVisibility.shared),
      ),
      isTrue,
    );
    expect(
      RouteVisibility.mayContributeRide(
        'saved-x',
        const SavedRouteMeta(catalogTourId: 'r-bodensee-road'),
      ),
      isTrue,
    );
  });

  test('shareHonesty mentions server revoke when logged in', () {
    final text = RouteVisibility.shareHonesty(
      routeId: 'gpx-1',
      hasTrack: true,
    );
    expect(text, contains('Server'));
    expect(text, contains('eingeloggt'));
  });

  test('catalog honesty avoids öffentlich and Akte', () {
    final text = RouteVisibility.shareHonesty(
      routeId: 'r-bodensee-road',
      hasTrack: false,
      meta: const SavedRouteMeta(catalogTourId: 'r-bodensee-road'),
    );
    expect(text, contains('freigegeben'));
    expect(text, isNot(contains('öffentlich')));
    expect(text, isNot(contains('Akte')));
  });

  test('collection share drops private GPX ids', () {
    expect(
      RouteVisibility.shareableRouteIds(
        ['gpx-1', 'gpx-2', 'r-bodensee-road'],
        {
          'gpx-2': const SavedRouteMeta(visibility: RouteVisibility.shared),
          'r-bodensee-road':
              const SavedRouteMeta(catalogTourId: 'r-bodensee-road'),
        },
      ),
      ['gpx-2', 'r-bodensee-road'],
    );
  });
}
